code/modules/research/ordnance/tank_compressor.dm

```diff
@@ -0,0 +1,330 @@
+#define TANK_COMPRESSOR_PRESSURE_LIMIT 5000
+#define TANK_COMPRESSOR_MAX_TRANSFER_RATE 20
+#define SIGNIFICANT_AMOUNT_OF_MOLES 10
+
+/obj/machinery/atmospherics/components/binary/tank_compressor
+	name = "Tank Compressor"
+	desc = "Heavy duty shielded air compressor designed to pressurize tanks above the safe limit."
+	circuit = /obj/item/circuitboard/machine/tank_compressor
+	icon = 'icons/obj/machines/research.dmi'
+	base_icon_state = "tank_compressor"
+	icon_state = "tank_compressor-open"
+	density = TRUE
+
+	var/active = FALSE
+	var/transfer_rate = TANK_COMPRESSOR_MAX_TRANSFER_RATE
+	var/datum/gas_mixture/leaked_gas_buffer
+	var/list/compressor_record
+	var/last_recorded_pressure = 0
+	var/record_number = 1
+	var/obj/item/tank/inserted_tank
+	/// Reference to a disk we are going to print to.
+	var/obj/item/computer_hardware/hard_drive/portable/inserted_disk
+
+	pipe_flags = PIPING_ONE_PER_TURF | PIPING_DEFAULT_LAYER_ONLY
+
+/obj/machinery/atmospherics/components/binary/tank_compressor/Initialize(mapload)
+	. = ..()
+	leaked_gas_buffer = new(200)
+	compressor_record = list()
+
+/obj/machinery/atmospherics/components/binary/tank_compressor/ComponentInitialize()
+	. = ..()
+	RegisterSignal(src, COMSIG_ATOM_INTERNAL_EXPLOSION, .proc/explosion_handle)
+
+/obj/machinery/atmospherics/components/binary/tank_compressor/examine()
+	. = ..()
+	. += "This one is rated for up to [TANK_COMPRESSOR_PRESSURE_LIMIT] kPa."
+
+/// Stores the record of the gas data for a significant enough tank leak
+/datum/data/compressor_record
+	/// Tank Name
+	var/experiment_source
+	/// Key: Path, Value: Moles
+	var/list/gas_data = list()
+	var/timestamp
+
+/obj/machinery/atmospherics/components/binary/tank_compressor/attackby(obj/item/item, mob/living/user)
+	if (!panel_open)
```

Comment:
early return?

Reply:
Some tool checks need to be impartial about the panel being open, while some need to make sure it is open. It's possible to do panel checks twice or add variables but this I reckon that this is tidier than those.

Reply:
Actually on second thought the tool checks are already scanning for the panel. I'll do it.